### ISSUE TITLE
Not using the entry in incremental build

### ISSUE THREAD

**Issue by zrii** [ngryman/gulp-bro#4]
The changes made in entry are not being implemented in the incremental build, only first time it builds it builds correctly. Every incremental will remain the same, no matter the changes in the initial entry.

**Comment by ngryman** [comment]
Hi,

I want to be sure I understand the bug correctly.

Context:

```javascript
gulp.task('build', () =>
  gulp.src('app.js')
    .pipe(bro())
    .pipe(gulp.dest('dist'))
)

gulp.watch('*.js', ['build'])
```

Do you mean that any change to the `app.js` file is not reflected in the bundle?

**Comment by zrii** [comment]
Exactly.

Changes in app.js just wont reflect on the bundle. Putting a simple alert wont be seen.
I added the cacheFile option to test, and I can say that the initial stream is cached and in the dependencies all.js is not to be found (among dependencies). But you will see as soon as you test.

**Comment by ngryman** [comment]
I confirm the bug, with the only difference being that build `n` reflect `n-1` changes.
At first glance it might be a [browserify-incremental](https://github.com/jsdf/browserify-incremental) issue. I'll investigate.
Thanks for reporting it.

**Comment by zrii** [comment]
I tested again to make sure, but in my example even n-1 wasn't reflected.
I keep adding changes and none are there no matter how many times I change.

Also I made my own setup with browserify-incremental and it's json file contains mtimes object, which indeed contains the stamp of my entry file. I guess there is the problem, since in bro build of cache file mtime is missing since entry is a stream.

I think if you bundle browserify directly to src file instead of its stream it will be fixed, but its a wild guess from my side.

**Comment by ngryman** [comment]
Have you tested with `gulp.src('whatever.js', { read: false })`?

**Comment by zrii** [comment]
This does the trick. Thanks!

**Comment by ngryman** [comment]
I think part of the problem might come from `browserify-cache-api`. I filed an issue here: https://github.com/jsdf/browserify-cache-api/issues/5.

**Comment by ngryman** [comment]
I got too quick on this one ;) That's not the problem.

**Comment by ngryman** [comment]
Everything should work fine with `{ read: false }`.

The `n-1` problem was a false one. My tests were too quick for `mtime` to be resolved. It can take up to 2 seconds depending on the `os`.

So the bug remains with streams.

**Comment by ngryman** [comment]
After some investigations, streams are not a problem. I had an issue when the file was empty, but I consider this as an edge case. Closing this.

**Comment by amberv0** [comment]
Hi! Thanks for this cool module!
Could you please update the documentation to include and explain this { read: false } thing? I just stumbled upon the very same problem, and it took me a while to arrive to this closed issue. 
Thanks!

**Comment by ngryman** [comment]
@amberv0 `{ read: false }` is part of `gulp` and only avoids reading file contents at the `gulp.src` level. It is not mandatory at all and should not have any consequence on how `bro` works.

If I remember well, this issue was only happening when the file was empty. Does `{ read: false }` resolves any issue for you?

**Comment by amberv0** [comment]
@ngryman, my issue was exactly the same as described in the very first post of this thread. The file was not empty, no.

**Comment by ngryman** [comment]
Ok then, I have to investigate again on this one, I don't remember it well. Do you confirm that with the latest version of **bro** the bug is still present if you omit `{ read: false }`?

**Comment by amberv0** [comment]
In my case it was 0.2.2. I guess it's the latest.

**Comment by ngryman** [comment]
I can confirm that this test does actually pass: https://github.com/ngryman/gulp-bro/blob/master/test/task.js#L117. I don't know why any exception is not thrown.

**Comment by amberv0** [comment]
I've set up the most basic example, here is the repo:
https://github.com/amberv0/bro-issue
run `gulp watch`
if you modify `index.js`, you'll see no changes in the bundle. If you modify `another.js`, everything is ok. Adding `read: false` removes the issue.

**Comment by XuluWarrior** [comment]
Would it be possible to put a warning in the readme.md when it talks about incremental builds.  I managed to waste an hour or two wondering what I was doing wrong.

**Comment by ngryman** [comment]
@XuluWarrior Yes I will mention it a FAQ section. I still don't have any response from `browserify-incremental`, which bothers me a little... I'll ping the author again, if not, I have to think of a workaround for this.

**Comment by XuluWarrior** [comment]
Thanks.  Hopefully the new FAQ section will save some confusion.

**Comment by ProLoser** [comment]
Would like to bump this issue. It looks like you closed https://github.com/jsdf/browserify-cache-api/issues/5 prematurely?

**Comment by ngryman** [comment]
@ProLoser Why did I closed this? That's a good question, I don't remember 👻
Another pending issue was https://github.com/jsdf/browserify-incremental/issues/33. Problem is, I don't have any answer from the author... I think I'm gonna patch it myself.

**Comment by ngryman** [comment]
I ended understanding what's happening here, at least a part of it 🎉 

`{ read: false }` tells **gulp** to not read the file contents but to delegate it to a task. More generally, I wanted to support streams as it allows **bro** to be more flexible (i.e other tasks can be invoked before pre-processing the entry content).

But supporting streams leads to a conflict with how **browserify-incremental** works. It basically rely on `mtime` of files that live on the filesystem. When specifying `{ read: false }`, it works fine as it handles the file itself. But when omitting it, and thus passing a stream, it loses the file path information and can't stat it to retrieve its `mtime`. The file never gets invalidated.

Two solutions here, either:

1. I drop stream support (ignoring contents read by **gulp**). It's a minimal change but we lose the ability to use other tasks before **bro**.

2. I patch **browserify-incremental** to support streams. I played a little with it and found the buggy spots. The stream entry file gets invalidated correctly, but I still don't manage to get the updated content in the bundle...

I'll keep you guys posted.
